tools/tpm2_ptool.py

```diff
@@ -1063,175 +1291,90 @@ def generate_options(self, group_parser):
-                    salt = sealobject['soauthsalt']
-                    iters = sealobject['soauthiters']
-                else:
-                    sealpub = sealobject['userpub']
-                    sealpriv = sealobject['userpriv']
-                    salt = sealobject['userauthsalt']
-                    iters = sealobject['userauthiters']
+        with TemporaryDirectory() as d:
+            tpm2 = Tpm2(d, path)
+            tertiarypriv, tertiarypub, tertiarypubdata = tpm2.importkey(sobjctx, sobjauth, objauth, privkey=privkey, alg=alg)
 
-                salt = binascii.unhexlify(salt)
-                sealauth = hash_pass(pin.encode(), iters, salt)['hash']
+            return (tertiarypriv, tertiarypub, tertiarypubdata)
 
-                # Load the so sealobject using the PARENTS AUTH (primaryobject)
-                sealctx = tpm2.load(pobj['handle'], pinpobjauth, sealpriv, sealpub)
+    def __call__(self, args):
+        super(self.__class__, self).__call__(args)
+        keylabel= args['key_label']
+        print('Imported key as label: "{keylabel}"'.format(keylabel=keylabel))
 
-                # Now that the sealobject is loaded, we need to unseal the wrapping key
-                # object auth or the key when the TPM doesn't support encryptdecrypt
 
-                wrappingauth = tpm2.unseal(sealctx, sealauth)
+@commandlet("addkey")
+class AddKeyCommand(NewKeyCommandBase):
+    '''
+    Adds a key to a token within a tpm2-pkcs11 store.
+    '''
 
-                sym_support = str2bool(token_config['sym-support'])
-                if sym_support:
-                    # Now that we have the unsealed wrappingauth value
-                    # Load the wrapping object
-                    wrappingobj = db.getwrapping(token['id'])
-                    wrappingctx = tpm2.load(pobj['handle'], pinpobjauth, wrappingobj['priv'], wrappingobj['pub'])
+    # adhere to an interface
+    # pylint: disable=no-self-use
+    def generate_options(self, group_parser):
+        group_parser.add_argument(
+            '--label',
+            help='The tokens label to add a key too.\n',
+            required=True)
+        group_parser.add_argument(
+            '--algorithm',
+            help='The type of the key.\n',
+            choices=[ 'rsa1024', 'rsa2048', 'aes128', 'aes256' ],
+            required=True)
+        group_parser.add_argument(
+            '--key-label',
+            help='The key label to identify the key. Defaults to an integer value')
+        group_parser.add_argument(
+            '--id',
+            help='The key id. Defaults to a random 8 bytes of hex.\n',
+            default=binascii.hexlify(os.urandom(8)))
+        pinopts = group_parser.add_mutually_exclusive_group(required=True)
+        pinopts.add_argument(
+            '--sopin',
+            help='The Administrator pin,'),
+        pinopts.add_argument(
+            '--userpin',
+            help='The User pin.\n'),
```

Comment:
both commandlets have similar options, the ones that are the same should be in the super class generate_options and then the subclass can call super, and then add the ones they need.

Reply:
looks like just a couple of the options are common, either the help string has slight modification or the options available are different. But anyway I'll make the change that you suggested even if its just 1 or 2 common options.

Reply:
Not python specific but an OOO language design pattern. If your parent class has function foo, and your child class has function foo then the childs routine is called when calling on type child. So if you want to perfrom what the parent class is doing and add to it, you call the parent class first and then add your logic. 

Here is some sample code:
```PY
#!/usr/bin/env python

import argparse

class parent(object):

    def __init__(self, subparser):
        self.subparser = subparser 

    def addoptions(self):
        s = self.subparser
        s.add_argument(
            '--common',
            type=int,
            help='common option'),

class child(parent):

    def addoptions(self):
        super(child, self).addoptions()

        s = self.subparser
        s.add_argument(
            '--child',
            type=int,
            help='child option'),


def main():
    
    opt_parser = argparse.ArgumentParser(
        description='A tool for manipulating the tpm2-pkcs11 database')

    subparsers = opt_parser.add_subparsers(help='commands')

    parent_subparser = subparsers.add_parser('parent')
    child_subparser = subparsers.add_parser('child')

    p = parent(parent_subparser)
    p.addoptions()
    
    c = child(child_subparser)
    c.addoptions()

    opt_parser.parse_args()

if __name__ == "__main__":
    main()
```

Reply:
Sure, this is not python related. I was going to implement it this way. I guess the python part question was for the above comment.